inc/password.inc.php

```diff
@@ -0,0 +1,149 @@
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ *  Password functions
+ *
+ * @package     Poweradmin
+ * @copyright   2007-2010 Rejo Zenger <rejo@zenger.nl>
+ * @copyright   2010-2015 Poweradmin Development Team
+ * @license     http://opensource.org/licenses/GPL-3.0 GPL
+ */
+namespace {
+
+    if (!function_exists('gen_mix_salt')) {
+        /**
+         * Generate random salt and salted password
+         *
+         * @param string $pass password
+         *
+         * @return salted password
+         */
+        function gen_mix_salt($pass) {
+            $salt = generate_salt();
+            return mix_salt($salt, $pass);
+        }
+
+        /**
+         * Generate salted password
+         *
+         * @param string $salt salt
+         * @param string $pass password
+         *
+         * @return string salted password
+         */
+        function mix_salt($salt, $pass) {
+            return md5($salt . $pass) . ':' . $salt;
+        }
+
+        /**
+         * Extract salt from password
+         *
+         * @param string $password salted password
+         *
+         * @return string salt
+         */
+        function extract_salt($password) {
+            return substr(strchr($password, ':'), 1);
+        }
+
+        /**
+         * Generate random salt for encryption
+         *
+         * @param int $len salt length (default=5)
+         *
+         * @return string salt string
+         */
+        function generate_salt($len = 5) {
+            $valid_characters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890@#$%^*()_-!';
+            $valid_len = strlen($valid_characters) - 1;
+            $salt = '';
+
+            for ($i = 0; $i < $len; $i++) {
+                $salt .= $valid_characters[mt_rand(0, $valid_len)];
+            }
+
+            return $salt;
+        }
+
+    }
+}
+
+namespace Poweradmin\password {
+
+    if (!function_exists('Poweradmin\\password\\hash')) {
+
+        function hash($password) {
+            global $password_encryption, $password_encryption_cost;
+
+            if ($password_encryption === 'md5salt') {
+                return gen_mix_salt($password);
+            } elseif ($password_encryption === 'bcrypt') {
+                return password_hash($password, PASSWORD_BCRYPT, ['cost' => $password_encryption_cost]);
+            } else {
+                return md5($password);
+            }
+        }
+
+        function verify($password, $hash) {
+            global $password_encryption;
+
+            if ($password_encryption === 'md5salt') {
+                return _strsafecmp(mix_salt(extract_salt($hash), $password), $hash);
+            } elseif ($password_encryption === 'bcrypt') {
+                return password_verify($password, $hash);
+            } else {
+                return _strsafecmp(md5($password), $hash);
+            }
+        }
+
+        function needs_rehash($hash) {
+            // @todo
```

Comment:
It would just be a wrapper around
http://php.net/manual/en/function.password-needs-rehash.php in case one
would want to upgrade his bcrypt settings, e.g. make sure that users after
successfully logging in have their password hashes updated.

It's also fairly simple to upgrade from md5 to bcrypt in the background
(e.g. check if hex + 32 chars).

I didn't implement it because I did not get any feedback for a long time,
and naturally there are some stuff which I wouldn't even want to make
backward compatible because they are so outdated :)